File: web/src/lib/api/memory.ts
import { request } from "./client";

export interface MemoryNode {
  id: string;
  node_type: string;
  scope_id: string;
  level: string;
  title: string;
  summary?: string | null;
  content: string;
  child_node_ids: string[];
  derived_from_notes: string[];
  derived_from_sources: string[];
  derived_from_chunks: number[];
  metadata_?: Record<string, unknown> | null;
  confidence_score?: number | null;
  created_at: string;
  updated_at: string;
}

export interface MemoryNodeList {
  items: MemoryNode[];
  total: number;
}

export interface MemoryEdge {
  id: number;
  user_id: string;
  from_node_id: string;
  to_kind: string;
  to_id: string;
  edge_type: string;
  weight?: number | null;
  description?: string | null;
  metadata_?: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
}

export interface MemoryEdgeList {
  items: MemoryEdge[];
  total: number;
}

export interface MemoryGraph {
  nodes: MemoryNode[];
  edges: MemoryEdge[];
}

export interface TopicMemoryCandidate {
  node: MemoryNode;
  reason: string;
}

export interface TopicMemoryCandidateList {
  items: TopicMemoryCandidate[];
  total: number;
}

export interface MemoryListParams {
  node_type?: string;
  level?: string;
  scope_id?: string;
  q?: string;
  status?: "active" | "archived" | "pending_review" | "rejected" | "merged";
  has_embedding?: boolean;
  stale?: boolean;
  source?: string;
  limit?: number;
  offset?: number;
}

export interface MemoryNodeUpdate {
  title?: string;
  summary?: string | null;
  content?: string;
  status?: "active" | "archived" | "rejected" | "merged";
}

export interface TopicMemoryCandidateParams {
  topic: string;
  limit?: number;
}

export interface TopicMemoryCompileRequest {
  topic: string;
  level?: "topic" | "subtopic";
  source_node_ids?: string[];
  limit?: number;
}

function toQuery(params: MemoryListParams | TopicMemoryCandidateParams = {}) {
  const search = new URLSearchParams();
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null && value !== "") search.set(key, String(value));
  });
  const query = search.toString();
  return query ? `?${query}` : "";
}

export const memoryApi = {
  list: (params?: MemoryListParams) => request<MemoryNodeList>(`/memory${toQuery(params)}`),
  get: (id: string) => request<MemoryNode>(`/memory/${encodeURIComponent(id)}`),
  update: (id: string, body: MemoryNodeUpdate) =>
    request<MemoryNode>(`/memory/${encodeURIComponent(id)}`, { method: "PATCH", body: JSON.stringify(body) }),
  edges: (id: string, params?: { direction?: "in" | "out" | "both"; limit?: number; refresh?: boolean }) => {
    const q = new URLSearchParams();
    if (params?.direction) q.set("direction", params.direction);
    if (params?.limit) q.set("limit", String(params.limit));
    if (params?.refresh !== undefined) q.set("refresh", String(params.refresh));
    return request<MemoryEdgeList>(`/memory/${encodeURIComponent(id)}/edges?${q}`);
  },
  graph: (id: string, params?: { limit?: number; refresh?: boolean }) => {
    const q = new URLSearchParams();
    if (params?.limit) q.set("limit", String(params.limit));
    if (params?.refresh !== undefined) q.set("refresh", String(params.refresh));
    return request<MemoryGraph>(`/memory/${encodeURIComponent(id)}/graph?${q}`);
  },
  topicCandidates: (params: TopicMemoryCandidateParams) =>
    request<TopicMemoryCandidateList>(`/memory/topics/candidates${toQuery(params)}`),
  compileTopic: (body: TopicMemoryCompileRequest) =>
    request<MemoryNode>("/memory/topics/compile", {
      method: "POST",
      body: JSON.stringify(body),
    }),
  merge: (id: string, body: { target_node_id: string; archive_source?: boolean }) =>
    request<MemoryNode>(`/memory/${encodeURIComponent(id)}/merge`, {
      method: "POST",
      body: JSON.stringify(body),
    }),
};

File: src/pkg/schemas/memory.py
from datetime import datetime

from pydantic import BaseModel, Field


class MemoryNodeRead(BaseModel):
    model_config = {"from_attributes": True}

    id: str
    node_type: str
    scope_id: str
    level: str
    title: str
    summary: str | None = None
    content: str
    child_node_ids: list[str]
    derived_from_notes: list[str]
    derived_from_sources: list[str]
    derived_from_chunks: list[int]
    metadata_: dict | None = Field(None, alias="metadata_")
    confidence_score: float | None = None
    created_at: datetime
    updated_at: datetime


class MemoryNodeList(BaseModel):
    items: list[MemoryNodeRead]
    total: int


class MemoryNodeUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=500)
    summary: str | None = None
    content: str | None = None
    status: str | None = Field(default=None, pattern=r"^(active|archived|rejected|merged)$")


class TopicMemoryCompileRequest(BaseModel):
    topic: str = Field(min_length=1, max_length=200)
    level: str = Field(default="topic", pattern=r"^(topic|subtopic)$")
    source_node_ids: list[str] | None = None
    limit: int = Field(default=20, ge=1, le=100)


class TopicMemoryCandidate(BaseModel):
    node: MemoryNodeRead
    reason: str


class TopicMemoryCandidateList(BaseModel):
    items: list[TopicMemoryCandidate]
    total: int


class MemoryEdgeRead(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    user_id: str
    from_node_id: str
    to_kind: str
    to_id: str
    edge_type: str
    weight: float | None = None
    description: str | None = None
    metadata_: dict | None = Field(None, alias="metadata_")
    created_at: datetime
    updated_at: datetime


class MemoryEdgeList(BaseModel):
    items: list[MemoryEdgeRead]
    total: int


class MemoryGraphRead(BaseModel):
    nodes: list[MemoryNodeRead]
    edges: list[MemoryEdgeRead]


class MemoryNodeMergeRequest(BaseModel):
    target_node_id: str = Field(min_length=1, max_length=300)
    archive_source: bool = True
